### PR TITLE
Re-Implement join throttle, stop packet processing after disconnect in InitialHandler

### PR DESCRIPTION
**Pull request by Janmm14**
This PR is created from [changes](https://github.com/minotopiame/FlexPipe/blob/master/patches/0004-Re-Implement-join-throttle.patch) done in [FlexPipe](https://github.com/minotopiame/FlexPipe).

This join throttle takes care of pings by kicking in only if its clearly a join try.
This implementation sends a kick message containing the seconds players needs to wait between join attempts.
Additionally it ignores any packets and further disconnects after the first disconnect was issued. ```ChannelWrapper.isClosed()``` is similar, however the close is delayed in ```InitialHandler.disconnect(...)``` and therefore we need to handle our own boolean.

Adapted stop of packet processing after disconnect from WaterfallMC/Waterfall@45e4a61 by @Techcable

**Edit:** First concerns answered, removed ```log_throttled``` config option.

**Edit 2:** Update ```ConnectionThrottle```: Use a concurrent guava cache with access expiration, so the throttle clears itself while also being *thread safe* now.

**Edit 3:** Connection throttle code is now in the switch below.

### CODE REVIEW

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java [hunk 1]
```diff
@@ -254,7 +261,21 @@ public void handle(Handshake handshake) throws Exception
         this.handshake = handshake;
         ch.setVersion( handshake.getProtocolVersion() );
 
-        // Starting with FML 1.8, a "\0FML\0" token is appended to the handshake. This interferes 
+        if ( handshake.getRequestedProtocol() == 2 && BungeeCord.getInstance().getConnectionThrottle().throttle( ( ( InetSocketAddress ) ch.getHandle().remoteAddress() ).getAddress() ) )
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: Magic value

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java [hunk 2]
```diff
@@ -254,7 +261,21 @@ public void handle(Handshake handshake) throws Exception
...
+        {
+            if ( BungeeCord.getInstance().getConfig().isLogThrottled() )
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: How is this configuration useful?

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java [hunk 3]
```diff
@@ -254,7 +261,21 @@ public void handle(Handshake handshake) throws Exception
...
+            {
+                bungee.getLogger().log( Level.INFO, "{0} throttled", this );
+            }
+            // setting thisState to username to stop possible code execution on repeated handshakes
+            thisState = State.USERNAME;
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: Make this portable; put in disconnect class.

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java [hunk 4]
```diff
@@ -285,9 +306,9 @@ public void handle(Handshake handshake) throws Exception
                 ch.setProtocol( Protocol.STATUS );
                 break;
             case 2:
+                // Login
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: Unrelated change

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java [hunk 5]
```diff
@@ -519,8 +540,9 @@ public void disconnect(String reason)
     @Override
     public void disconnect(final BaseComponent... reason)
     {
-        if ( !ch.isClosed() )
+        if ( !disconnecting || !ch.isClosed() )
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: See implementation of isClosed

File: proxy/src/main/resources/messages.properties [hunk 6]
```diff
@@ -24,3 +24,5 @@ total_players=Total players online: {0}
 name_too_long=Cannot have username longer than 16 characters
 name_invalid=Username contains invalid characters.
 ping_cannot_connect=\u00a7c[Bungee] Can't connect to server.
+
```
Comment (md-5) [inline proxy/src/main/resources/messages.properties]: Erroneous newline

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java [hunk 7]
```diff
@@ -98,6 +99,13 @@ public void sendPacket(DefinedPacket packet)
     private boolean legacy;
     @Getter
     private String extraDataInHandshake = "";
+    private boolean disconnecting = false;
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: See comments re: isClosed

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java [hunk 8]
```diff
@@ -213,7 +221,6 @@ public void done(ServerPing result, Throwable error)
                     @Override
                     public void done(ProxyPingEvent pingResult, Throwable error)
                     {
-                        BungeeCord.getInstance().getConnectionThrottle().unthrottle( getAddress().getAddress() );
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: You're never unthrottling, this will break Minecraft versions which double ping,

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 1 above)
Comment (Janmm14) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: As magic as in the switch below. Should I add another comment here that the two means?

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 2 above)
Comment (Janmm14) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: Will remove.

File: api/src/main/java/net/md_5/bungee/api/ProxyConfig.java [hunk 9]
```diff
@@ -46,6 +46,11 @@
     boolean isLogCommands();
 
     /**
+     * Whether we log throttled join attempts
+     */
+    boolean isLogThrottled();
```
Comment (md-5) [inline api/src/main/java/net/md_5/bungee/api/ProxyConfig.java]: See below discussion, this seems useless.

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 4 above)
Comment (Janmm14) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: If you think this should be taken out, I can do that.

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 3 above)
Comment (Janmm14) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: This is just equal to "accepting" the handshake as done in the switch below. I don't see your point here.

Or are you referring to the lines 272-273? I can include [FlexPipe patch 34](https://github.com/minotopiame/FlexPipe/blob/master/patches/0034-Ensure-kick-messages-can-be-sent-while-being-at-prot.patch) also if you want.

(The only reason this is at the top is to not execute the code until the switch.)

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 7 above)
Comment (Janmm14) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: The problem is that the channel close is delayed 500ms, in this time we would then recieve the users ```LoginRequest``` (as its send together).

**Comment by Janmm14** [comment]
Btw if you think its ready, let me know and I can squash the commits if you want.

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 7 above)
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: No.
https://github.com/SpigotMC/BungeeCord/blob/master/proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java#L59

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 1 above)
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: Then put this code in the switch below?

File: proxy/src/main/java/net/md_5/bungee/ConnectionThrottle.java [hunk 10]
```diff
@@ -1,25 +1,30 @@
 package net.md_5.bungee;
 
 import java.net.InetAddress;
-import java.util.HashMap;
-import java.util.Map;
 import lombok.RequiredArgsConstructor;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.util.concurrent.TimeUnit;
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/ConnectionThrottle.java]: Bad import ordering.

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 7 above)
Comment (Janmm14) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: Yes.
https://github.com/SpigotMC/BungeeCord/blob/master/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java#L529-541

File: proxy/src/main/java/net/md_5/bungee/ConnectionThrottle.java (hunk 10 above)
Comment (Janmm14) [inline proxy/src/main/java/net/md_5/bungee/ConnectionThrottle.java]: Fixed now with alphabetical order.

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 1 above)
Comment (Janmm14) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: Done.

File: proxy/src/main/resources/messages.properties (hunk 6 above)
Comment (Janmm14) [inline proxy/src/main/resources/messages.properties]: Is removed.

**Comment by Janmm14** [comment]
@md-5 any answer?

**Comment by Janmm14** [comment]
Stil waiting for a reply, is this now ready or not? Do you want that I squash the commits now?

Please answer.

**Comment by md-5** [comment]
2e8ed1cfba4a23ee0c6aa723404482a9889eb040

**Comment by Janmm14** [comment]
Thanks.
